addons/project/models/project.py

```diff
@@ -367,13 +369,18 @@ def unlink(self):
         return result
 
     def message_subscribe(self, partner_ids=None, channel_ids=None, subtype_ids=None):
-        """ Subscribe to all existing active tasks when subscribing to a project """
+        """
+        Subscribe to all existing active tasks when subscribing to a project
+        And add the portal user subscribed to allowed portal users
+        """
         res = super(Project, self).message_subscribe(partner_ids=partner_ids, channel_ids=channel_ids, subtype_ids=subtype_ids)
         project_subtypes = self.env['mail.message.subtype'].browse(subtype_ids) if subtype_ids else None
         task_subtypes = project_subtypes.mapped('parent_id').ids if project_subtypes else None
         if not subtype_ids or task_subtypes:
             self.mapped('tasks').message_subscribe(
                 partner_ids=partner_ids, channel_ids=channel_ids, subtype_ids=task_subtypes)
+        for user_id in self.env['res.partner'].browse(partner_ids).mapped('user_ids').filtered('share'):
+            self.allowed_portal_user_ids |= user_id
```

Comment:
`self.allowed_portal_user_ids |=  self.env['res.partner'].browse(partner_ids).user_ids.filtered('share')`
no for loop needed ;)